// @mysten/walrus/dist/cjs/quilt/write.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var write_exports = {};
__export(write_exports, {
  encodeQuilt: () => encodeQuilt
});
module.exports = __toCommonJS(write_exports);
var import_bcs = require("@mysten/bcs");
var import_bcs2 = require("../utils/bcs.js");
var import_utils = require("../utils/index.js");
var import_quilts = require("../utils/quilts.js");
function encodeQuilt({ blobs, numShards, encodingType }) {
  const { primarySymbols: nRows, secondarySymbols: nCols } = (0, import_utils.getSourceSymbols)(
    numShards,
    encodingType
  );
  const sortedBlobs = blobs.sort((a, b) => a.identifier < b.identifier ? -1 : 1);
  const identifiers = /* @__PURE__ */ new Set();
  const index = {
    patches: []
  };
  const tags = sortedBlobs.map(
    (blob) => blob.tags && Object.keys(blob.tags).length > 0 ? import_bcs2.QuiltPatchTags.serialize(blob.tags).toBytes() : null
  );
  for (const blob of sortedBlobs) {
    if (identifiers.has(blob.identifier)) {
      throw new Error(`Duplicate blob identifier: ${blob.identifier}`);
    }
    identifiers.add(blob.identifier);
    index.patches.push({
      startIndex: 0,
      endIndex: 0,
      identifier: blob.identifier,
      tags: blob.tags ?? {}
    });
  }
  const indexSize = import_quilts.QUILT_INDEX_PREFIX_SIZE + import_bcs2.QuiltIndexV1.serialize(index).toBytes().length;
  const blobMetadata = sortedBlobs.map((blob, i) => {
    const identifierBytes = import_bcs.bcs.string().serialize(blob.identifier).toBytes();
    let metadataSize = import_quilts.QUILT_PATCH_BLOB_HEADER_SIZE + import_quilts.BLOB_IDENTIFIER_SIZE_BYTES_LENGTH + identifierBytes.length;
    let mask = 0;
    let offset = 0;
    if (tags[i]) {
      metadataSize += import_quilts.TAGS_SIZE_BYTES_LENGTH + tags[i].length;
      mask |= import_quilts.HAS_TAGS_FLAG << 0;
    }
    const metadata = new Uint8Array(metadataSize);
    const metadataView = new DataView(metadata.buffer);
    const header = import_bcs2.QuiltPatchBlobHeader.serialize({
      version: 1,
      length: metadataSize - import_quilts.QUILT_PATCH_BLOB_HEADER_SIZE + blob.contents.length,
      mask
    }).toBytes();
    metadata.set(header, offset);
    offset += header.length;
    metadataView.setUint16(offset, identifierBytes.length, true);
    offset += import_quilts.BLOB_IDENTIFIER_SIZE_BYTES_LENGTH;
    metadata.set(identifierBytes, offset);
    offset += identifierBytes.length;
    if (tags[i]) {
      metadataView.setUint16(offset, tags[i].length, true);
      offset += import_quilts.TAGS_SIZE_BYTES_LENGTH;
      metadata.set(tags[i], offset);
      offset += tags[i].length;
    }
    return metadata;
  });
  const blobSizes = [
    indexSize,
    ...sortedBlobs.map((blob, i) => {
      if (blob.identifier.length > import_quilts.MAX_BLOB_IDENTIFIER_BYTES_LENGTH) {
        throw new Error(`Blob identifier too long: ${blob.identifier}`);
      }
      return blobMetadata[i].length + blob.contents.length;
    })
  ];
  const symbolSize = (0, import_quilts.computeSymbolSize)(
    blobSizes,
    nCols,
    nRows,
    import_quilts.MAX_NUM_SLIVERS_FOR_QUILT_INDEX,
    encodingType
  );
  const rowSize = symbolSize * nCols;
  const columnSize = symbolSize * nRows;
  const indexColumnsNeeded = Math.ceil(indexSize / columnSize);
  if (indexColumnsNeeded > import_quilts.MAX_NUM_SLIVERS_FOR_QUILT_INDEX) {
    throw new Error("Index too large");
  }
  const quilt = new Uint8Array(rowSize * nRows);
  let currentColumn = indexColumnsNeeded;
  for (let i = 0; i < sortedBlobs.length; i++) {
    const blob = sortedBlobs[i];
    index.patches[i].startIndex = currentColumn;
    currentColumn += writeBlobToQuilt(
      quilt,
      blob.contents,
      rowSize,
      columnSize,
      symbolSize,
      currentColumn,
      blobMetadata[i]
    );
    index.patches[i].endIndex = currentColumn;
  }
  const indexBytes = import_bcs2.QuiltIndexV1.serialize(index).toBytes();
  const quiltIndex = new Uint8Array(import_quilts.QUILT_INDEX_PREFIX_SIZE + indexBytes.length);
  quiltIndex.set([1], 0);
  quiltIndex.set(new Uint32Array([indexBytes.length]), 1);
  quiltIndex.set(indexBytes, import_quilts.QUILT_INDEX_PREFIX_SIZE);
  writeBlobToQuilt(quilt, quiltIndex, rowSize, columnSize, symbolSize, 0);
  return { quilt, index };
}
function writeBlobToQuilt(quilt, blob, rowSize, columnSize, symbolSize, startColumn, prefix) {
  const nRows = rowSize / symbolSize;
  let bytesWritten = 0;
  if (rowSize % symbolSize !== 0) {
    throw new Error("Row size must be divisible by symbol size");
  }
  if (columnSize % symbolSize !== 0) {
    throw new Error("Column size must be divisible by symbol size");
  }
  if (prefix) {
    writeBytes(prefix);
  }
  writeBytes(blob);
  return Math.ceil(bytesWritten / columnSize);
  function writeBytes(bytes) {
    const offset = bytesWritten;
    const symbolsToSkip = Math.floor(offset / symbolSize);
    let remainingOffset = offset % symbolSize;
    let currentCol = startColumn + Math.floor(symbolsToSkip / nRows);
    let currentRow = symbolsToSkip % nRows;
    let index = 0;
    while (index < bytes.length) {
      const baseIndex = currentRow * rowSize + currentCol * symbolSize;
      const startIndex = baseIndex + remainingOffset;
      const len = Math.min(symbolSize - remainingOffset, bytes.length - index);
      const subarray = bytes.subarray(index, index + len);
      quilt.set(subarray, startIndex);
      index += len;
      remainingOffset = 0;
      currentRow = (currentRow + 1) % nRows;
      if (currentRow === 0) {
        currentCol++;
      }
    }
    bytesWritten += bytes.length;
  }
}
