// @mysten/walrus/dist/cjs/contracts/wal/wal.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var wal_exports = {};
__export(wal_exports, {
  ProtectedTreasury: () => ProtectedTreasury,
  TreasuryCapKey: () => TreasuryCapKey,
  WAL: () => WAL,
  burn: () => burn,
  totalSupply: () => totalSupply
});
module.exports = __toCommonJS(wal_exports);
var import_bcs = require("@mysten/sui/bcs");
var import_utils = require("../utils/index.js");
var object = __toESM(require("./deps/sui/object.js"));
function WAL() {
  return import_bcs.bcs.struct("WAL", {
    dummy_field: import_bcs.bcs.bool()
  });
}
function ProtectedTreasury() {
  return import_bcs.bcs.struct("ProtectedTreasury", {
    id: object.UID()
  });
}
function TreasuryCapKey() {
  return import_bcs.bcs.struct("TreasuryCapKey", {
    dummy_field: import_bcs.bcs.bool()
  });
}
function totalSupply(options) {
  const packageAddress = options.package ?? "@local-pkg/wal";
  const argumentsTypes = [`${packageAddress}::wal::ProtectedTreasury`];
  const parameterNames = ["treasury"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "wal",
    function: "total_supply",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
function burn(options) {
  const packageAddress = options.package ?? "@local-pkg/wal";
  const argumentsTypes = [
    `${packageAddress}::wal::ProtectedTreasury`,
    `0x0000000000000000000000000000000000000000000000000000000000000002::coin::Coin<${packageAddress}::wal::WAL>`
  ];
  const parameterNames = ["treasury", "coin"];
  return (tx) => tx.moveCall({
    package: packageAddress,
    module: "wal",
    function: "burn",
    arguments: (0, import_utils.normalizeMoveArguments)(options.arguments, argumentsTypes, parameterNames)
  });
}
